src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_QueryLambdaRewriter.vb

```diff
@@ -0,0 +1,93 @@
+
+Imports Microsoft.CodeAnalysis.VisualBasic
+Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+Namespace Microsoft.CodeAnalysis.Semantics
+    Partial Friend NotInheritable Class VisualBasicOperationFactory
+        Private Shared Function RewriteQueryLambda(node As BoundQueryLambda) As BoundNode
+            ' We rewrite query lambda into regular lambda with 2 passes.
+            ' Pass 1 uses helper methods from LocalRewriter to do the lowering. This introduces large number of DAGs.
+            ' Pass 2 walks over the lowered tree and replaces duplicate bound nodes in the tree with their clones - this is a requirement for the Operation tree.
+            Dim pass1Rewriter As New QueryLambdaRewriterPass1
+            Dim rewrittenLambda As BoundLambda = DirectCast(pass1Rewriter.VisitQueryLambda(node), BoundLambda)
+
+            Dim pass2Rewriter As New QueryLambdaRewriterPass2
+            Return pass2Rewriter.VisitLambda(rewrittenLambda)
+        End Function
+
+        Private NotInheritable Class QueryLambdaRewriterPass1
+            Inherits BoundTreeRewriterWithStackGuard
+            Private _rangeVariableMap As Dictionary(Of RangeVariableSymbol, BoundExpression)
+
+            Public Sub New()
+                _rangeVariableMap = Nothing
+            End Sub
+
+            Public Overrides Function VisitQueryLambda(node As BoundQueryLambda) As BoundNode
+                LocalRewriter.PopulateRangeVariableMapForQueryLambdaRewrite(node, _rangeVariableMap, inExpressionLambda:=False)
+                Dim rewrittenBody As BoundStatement = LocalRewriter.RewriteQueryLambdaBody(AddressOf VisitExpressionWithStackGuard, node, _rangeVariableMap)
+                Return LocalRewriter.RewriteQueryLambda(rewrittenBody, node)
+            End Function
+
+            Public Overrides Function VisitRangeVariable(node As BoundRangeVariable) As BoundNode
+                Dim expression As BoundExpression = Nothing
+                If Not _rangeVariableMap.TryGetValue(node.RangeVariable, expression) Then
+                    ' _rangeVariableMap should contain an entry for the range variable, except for error cases. 
+                    Debug.Assert(node.HasErrors OrElse node.RangeVariable.Type.IsErrorType())
+                    Return node
+                End If
+
+#If DEBUG Then
+                ' Range variable reference should be rewritten to a parameter reference, or a call or a property access.
+                ' We clone these bound nodes in QueryLambdaRewriterPass2 to avoid dag in the generated bound tree.
+                ' If the LocalRewriter is changed to generate more kind of bound nodes for range variables, we should handle these in QueryLambdaRewriterPass2.
+                ' Below assert helps us to stay in sync with the LocalRewriter.
+                Select Case expression.Kind
+                    Case BoundKind.Parameter
+                    Case BoundKind.Call
```

Comment:
>Case BoundKind.Call [](start = 20, length = 19)

I think for IOperation we should always use a property access instead of a call.